### PR TITLE
connection router to accomodate multithreading

### PR DESCRIPTION
**Pull request by CoryGH**
connection router hook added to accommodate multithreaded servers

e.g. passing `options.connectionRouter(connection)` will allow for:

    function connectionRouter(connection) {
        sendToThread(findRoute(connection), connection);
    }
    
    function sendToThread(threadId, connection) {
        ...wrap in packet for IPC and send to thread...
    }
    
    function threadMessage(message) {
        notListeningServerOnChildThread.newConnection(message.connection);
    }

in order to preserve extant functionality while adding multi-threading support.

### CODE REVIEW

**Review by UziTech — CHANGES_REQUESTED** [review]
Can you add a test and update docs?

**Comment by CoryGH** [comment]
> 
> 
> Can you add a test and update docs?

Sure, I think I have a handle on your test system.

**Comment by UziTech** [comment]
it seems like multi threading with the `connectionRouter` option would only make use of ldapjs to start the server.

It seems like it might be easier to start the server and send connection to worker threads without ldapjs.

```js
const net = require("net");
net.createServer(connectionRouter)
```

What are the benefits of using ldapjs to start the server when you are using `notListeningServerOnChildThread` to do all of the actual work?

**Comment by CoryGH** [comment]
> What are the benefits of using ldapjs to start the server when you are using `notListeningServerOnChildThread` to do all of the actual ldap work?

My aim was to make things minimally-invasive from an API standpoint.  The `connectionRouter` function is called at the master thread level to tell the thread-manager (whichever it may be) to forward to a child thread which must then call `newConnection(connection)` in order to initialize the ldapjs side for the connection.  It's designed around a forking design pattern but is extensible to cross-machine threading engines as well.

The `server.listen` function looks lightweight enough to do what you're suggesting (if what you mean is to create a net/tls server outside of ldapjs then use `server.newConnection(connection)` to leverage the ldapjs functionality with the socket (the change to make `server.newConnection(connection)` publicly-visible would still be needed, the `connectionRouter` piece wouldn't be needed at that point.  Generally I shy away from moving entire members outside of libraries wholesale like that (in this case reinforced when I was reviewing `server.newConnection(connection)` because there's a lot of library-specific code in there leveraging the this/self reference to set the appropriate state and there's no real guarantee some of that logic or things similar to it from the patterns in the code won't get moved to the `listen` function in the future (that seems like a bigger commitment to me on the part of library developers/maintainers than adding a hook, though I could be wrong.)

**Comment by CoryGH** [comment]
This is a mess in trying to get around the tap test framework's thread handling, here's what I have thus far though - might need to take a break from it today but the code paths are being hit (failing test with "test unfinished" and I haven't pinned down why as of yet (I placed this in `server.test.js`, had to wrap the other tests in `if (cluster.isMaster) { ... }` as well to avoid them running on child threads.

    tap.test('multithreading', function (t) {
      let threads = []
      const server = ldap.createServer({
        connectionRouter: (connection) => {
          if (cluster.isMaster) {
            threads[0].send({
              type: 'connection'
            }, connection)
          }
        }
      })
      if (cluster.isMaster) {
        let clientHolder = { client: null }
        for (let i = 0; i < 1; i++) {
          let thread = cluster.fork({
            id: i
          })
          thread.on('message', function (msg, connection) {
            switch (msg.type) {
              case 'connected':
                clientHolder.client.unbind()
                threads[0].disconnect()
                t.ok(true, 'client connected')
                t.end()
                server.close(() => t.end())
                break
            }
          })
          threads.push(thread)
        }
        t.ok(server)
        server.listen(5555, 'localhost', function () {
          t.ok(true, 'server listening on ' + server.url)
    
          server.getConnections(function (err, count) {
            t.error(err)
            t.equal(count, 0)
            setTimeout(() => {
              clientHolder.client = ldap.createClient({ port: 5555, url: server.url })
              clientHolder.client.on('connect', function () { })
            }, 1000)
          })
        })
      } else {
        process.on('message', (msg, connection) => {
          switch (msg.type) {
            case 'connection':
              server.newConnection(connection)
              process.send({
                type: 'connected'
              })
              break
          }
        })
      }
    })

**Comment by CoryGH** [comment]
I would agree that if the `server.listen()` call can be guaranteed not to include state-specific bindings as the library evolves it would be better to just make `server.newConnection(connection)` publicly-visible without the `connectionRouter(connection)` hook.

**Comment by UziTech** [comment]
I think it would be easiest to mock the threads instead of trying to run the tests in actual threads. Just testing that `connectionRouter` option does what it is supposed to and that `newConnection` is public should work.

**Comment by CoryGH** [comment]
> 
> 
> I think it would be easiest to mock the threads instead of trying to run the tests in actual threads. Just testing that `connectionRouter` option is called when it is supposed to be and that `newConnection` is public should work.

Would you like it with or without `connectionRouter` given the ability to bypass the listen function?

**Comment by UziTech** [comment]
> Would you like it with or without `connectionRouter` given the ability to bypass the listen function?

Maybe two tests, on each way?

**Comment by CoryGH** [comment]
> 
> 
> > Would you like it with or without `connectionRouter` given the ability to bypass the listen function?
> 
> Maybe two tests, one each way?

Tests added and connectionRouter hook added back in.

**Comment by CoryGH** [comment]
@UziTech updated to fix linting issue

File: lib/server.js [hunk 1]
```diff
@@ -313,7 +313,7 @@ function Server (options) {
     return c
   }
 
-  function newConnection (conn) {
+  self.newConnection = function (conn) {
```
Comment (jsumners) [inline lib/server.js:316]: Why does this need to be `self.newConnection`?

**Review by jsumners — COMMENTED** [review]
Are you able to add some documentation. An example in the docs would be quite helpful.

File: lib/server.js (hunk 1 above)
Comment (UziTech) [reply lib/server.js:316]: It just needs to be a public function. We should probably move it to the prototype.

Comment (CoryGH) [reply lib/server.js:316]: I can throw together a simple example for the documentation.

**Comment by CoryGH** [comment]
Added examples in the examples directory for both multithreading via the `connectionRouter` hook in conjunction with `newConnection(connection)` and another by replacing the whole listen server to use `newConnection(connection)` for the node-ldapjs bindings.

File: docs/examples.md [hunk 2]
```diff
@@ -550,3 +550,76 @@ To test out this example, try:
 $ ldapsearch -H ldap://localhost:389 -x -D cn=demo,dc=example,dc=com \
   -w demo -b "dc=example,dc=com" objectclass=*
 ```
+
+# Multi-threaded Server
+
+This example demonstrates multi-threading via the `cluster` module utilizing a `net` server for initial socket receipt.  An alternate example demonstrating use of the `connectionRouter` `serverOptions` hook is available in the `examples` directory.
+
+```
```
Comment (jsumners) [inline docs/examples.md:558]:
```suggestion
```js
```

File: lib/server.js [hunk 3]
```diff
@@ -313,7 +313,7 @@ function Server (options) {
...
     setupConnection(conn)
```
Comment (jsumners) [inline lib/server.js:317]:
```suggestion
    // TODO: make `newConnection` available on the `Server` prototype
    // https://github.com/ldapjs/node-ldapjs/pull/727/files#r636572294
    setupConnection(conn)
```

We'll solve this at a later date.

**Review by jsumners — APPROVED** [review]
Looks good to me once CI is green.

**Comment by UziTech** [comment]
@CoryGH you can run `npm run lint` to fix the linting issues.

**Comment by jsumners** [comment]
⚠️ This issue has been locked due to age. If you have encountered a recent
problem that seems to be covered by this issue, please open a new issue.

Please include a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)
when opening a new issue.

**Pull request merged**